kinto/authorization.py

```diff
@@ -94,40 +92,35 @@ def _resource_endpoint(object_uri):
     the specified `object_uri`. Returns `(None, None)` for the root URL plural
     endpoint.
     """
-    url_patterns = [
-        ('record', r'/buckets/(.+)/collections/(.+)/records/(.+)?'),
-        ('collection', r'/buckets/(.+)/collections/(.+)?'),
-        ('group', r'/buckets/(.+)/groups/(.+)?'),
-        ('bucket', r'/buckets/(.+)?'),
-        ('', r'/(buckets)')  # Root buckets list.
-    ]
-    for resource_name, pattern in url_patterns:
-        m = re.match(pattern, object_uri)
-        if m:
-            plural = '/' in m.groups()[-1]
-            return resource_name, plural
-    raise ValueError("%r is not a resource." % object_uri)
+    obj_parts = object_uri.split('/')
+    plural_endpoint = len(obj_parts) % 2 == 0
+    if plural_endpoint:
+        # /buckets/bid/collections -> /buckets/bid
+        obj_parts = obj_parts[:-1]
+
+    if len(obj_parts) <= 2:
+        # Root URL /buckets -> ('', False)
+        return '', False
+
+    # /buckets/bid -> buckets
+    resource_name = obj_parts[-2]
+    # buckets -> bucket
+    resource_name = resource_name.rstrip('s')
+    return resource_name, plural_endpoint
```

Comment:
Here basically the code does the same, but is not limited to particular patterns.